File: wrapper/src/ro_mol.cc
#include "rust/cxx.h"
#include <GraphMol/GraphMol.h>
#include <GraphMol/SmilesParse/SmilesParse.h>
#include <GraphMol/SmilesParse/SmilesWrite.h>
#include <DataStructs/ExplicitBitVect.h>
#include <GraphMol/Fingerprints/Fingerprints.h>
#include <GraphMol/MolStandardize/Tautomer.h>
#include <GraphMol/MolOps.h>

#include <iostream>

namespace RDKit {
    using ExplicitBitVect = ::ExplicitBitVect;

    std::shared_ptr<ROMol> copy_mol(const std::shared_ptr<ROMol> &mol) {
        return std::shared_ptr<ROMol>(new ROMol(*mol));
    }

    std::shared_ptr<ROMol> smiles_to_mol(const std::string &smiles) {
        ROMol *mol = SmilesToMol(smiles);

        return std::shared_ptr<ROMol>(mol);
    }

    rust::String mol_to_smiles(const std::shared_ptr<ROMol> &mol) {
        return MolToSmiles(*mol);
    }

    std::shared_ptr<ROMol> smiles_to_mol_with_params(const std::string &smiles, const std::shared_ptr<SmilesParserParams> &params) {
        ROMol *mol = SmilesToMol(smiles, *params);

        return std::shared_ptr<ROMol>(mol);
    }
    std::shared_ptr<SmilesParserParams> new_smiles_parser_params() {
        return std::shared_ptr<SmilesParserParams>(new SmilesParserParams());
    }
    void smiles_parser_params_set_sanitize(const std::shared_ptr<SmilesParserParams> &params, bool sanitize) {
        params->sanitize = sanitize;
    }



    using MolSanitizeExceptionUniquePtr = std::unique_ptr<MolSanitizeException>;
    std::unique_ptr<std::vector<MolSanitizeExceptionUniquePtr>> detect_chemistry_problems(const std::shared_ptr<ROMol> &mol) {
        std::vector<MolSanitizeExceptionUniquePtr> exceptions = RDKit::MolOps::detectChemistryProblems(*mol);
        std::vector<MolSanitizeExceptionUniquePtr> *heaped_exceptions = new std::vector<MolSanitizeExceptionUniquePtr>();

        auto s = exceptions.size();
        heaped_exceptions->reserve(s);
        for (int i=0; i<s; i++) {
          heaped_exceptions->push_back(std::move(exceptions[i]));
        }

        return std::unique_ptr<std::vector<MolSanitizeExceptionUniquePtr>>(heaped_exceptions);
    }

    rust::String mol_sanitize_exception_type(const MolSanitizeExceptionUniquePtr &mol_except) {
      return mol_except->getType();
    }

    unsigned int atom_sanitize_exception_get_atom_idx(const MolSanitizeExceptionUniquePtr &mol_except) {
      MolSanitizeException *mol_except_ptr = mol_except.get();
      AtomSanitizeException *atom_sanitize_except_ptr = (AtomSanitizeException *) mol_except_ptr;

      return atom_sanitize_except_ptr->getAtomIdx();
    }

    unsigned int get_num_atoms(const std::shared_ptr<ROMol> &mol, bool only_explicit) {
      return mol->getNumAtoms(only_explicit);
    }

    std::shared_ptr<Atom> get_atom_with_idx(const std::shared_ptr<ROMol> &mol, unsigned int idx) {
      Atom *borrowed_atom = mol->getAtomWithIdx(idx);
      Atom *atom = new Atom((*borrowed_atom));

      return std::shared_ptr<Atom>(atom);
    }

    rust::String get_symbol(const std::shared_ptr<Atom> &atom) {
      return atom->getSymbol();
    }

    bool get_is_aromatic(const std::shared_ptr<Atom> &atom) {
      return atom->getIsAromatic();
    }
    int get_atomic_num(const std::shared_ptr<Atom> &atom) {
      return atom->getAtomicNum();
    }
    int get_formal_charge(const std::shared_ptr<Atom> &atom) {
      return atom->getFormalCharge();
    }
    unsigned int get_total_num_hs(const std::shared_ptr<Atom> &atom) {
      return atom->getTotalNumHs();
    }
    unsigned int get_total_valence(const std::shared_ptr<Atom> &atom) {
      return atom->getTotalValence();
    }
    void set_formal_charge(std::shared_ptr<Atom> &atom, int what) {
      atom->setFormalCharge(what);
    }
    void set_num_explicit_hs(std::shared_ptr<Atom> &atom, int what) {
      atom->setNumExplicitHs(what);
    }
    void atom_update_property_cache(std::shared_ptr<Atom> &atom, bool strict) {
      atom->updatePropertyCache(strict);
    }

    void ro_mol_update_property_cache(std::shared_ptr<ROMol> &mol, bool strict) {
      mol->updatePropertyCache(strict);
    }
}

File: src/bridge/ro_mol.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");

        pub type ROMol;
        pub type ExplicitBitVect = crate::fingerprint_ffi::ExplicitBitVect;
        pub type SmilesParserParams;
        pub type Atom;

        pub type MolSanitizeException;
        pub type MolSanitizeExceptionUniquePtr; //  = UniquePtr<MolSanitizeException>;

        pub fn copy_mol(mol: &SharedPtr<ROMol>) -> SharedPtr<ROMol>;

        pub fn smiles_to_mol(smi: &CxxString) -> Result<SharedPtr<ROMol>>;

        pub fn smiles_to_mol_with_params(
            smi: &CxxString,
            params: &SharedPtr<SmilesParserParams>,
        ) -> Result<SharedPtr<ROMol>>;
        pub fn new_smiles_parser_params() -> SharedPtr<SmilesParserParams>;
        pub fn smiles_parser_params_set_sanitize(
            ptr: &SharedPtr<SmilesParserParams>,
            sanitize: bool,
        );

        pub fn mol_to_smiles(mol: &SharedPtr<ROMol>) -> String;

        pub fn detect_chemistry_problems(
            mol: &SharedPtr<ROMol>,
        ) -> UniquePtr<CxxVector<MolSanitizeExceptionUniquePtr>>;

        pub fn mol_sanitize_exception_type(
            mol_sanitize_exception: &MolSanitizeExceptionUniquePtr,
        ) -> String;

        pub fn atom_sanitize_exception_get_atom_idx(
            mol_sanitize_exception: &MolSanitizeExceptionUniquePtr,
        ) -> u32;

        pub fn get_num_atoms(mol: &SharedPtr<ROMol>, onlyExplicit: bool) -> u32;
        pub fn get_atom_with_idx(mol: &SharedPtr<ROMol>, idx: u32) -> SharedPtr<Atom>;
        pub fn get_symbol(atom: &SharedPtr<Atom>) -> String;
        pub fn get_is_aromatic(atom: &SharedPtr<Atom>) -> bool;
        pub fn get_atomic_num(atom: &SharedPtr<Atom>) -> i32;
        pub fn get_formal_charge(atom: &SharedPtr<Atom>) -> i32;
        pub fn get_total_num_hs(atom: &SharedPtr<Atom>) -> u32;
        pub fn get_total_valence(atom: &SharedPtr<Atom>) -> u32;
        pub fn set_formal_charge(atom: &mut SharedPtr<Atom>, what: i32);
        pub fn set_num_explicit_hs(atom: &mut SharedPtr<Atom>, what: i32);
        pub fn atom_update_property_cache(atom: &mut SharedPtr<Atom>, strict: bool);

        pub fn ro_mol_update_property_cache(atom: &mut SharedPtr<ROMol>, strict: bool);
    }
}
